autodispose/src/main/java/com/uber/autodispose/AutoDisposingCompletableObserverImpl.java

```diff
@@ -95,14 +91,16 @@ void callMainSubscribeIfNecessary(Disposable d) {
 
   @Override public void onComplete() {
     if (!isDisposed()) {
-      lazyDispose();
+      AutoDisposableHelper.dispose(lifecycleDisposable);
+      mainDisposable.lazySet(AutoDisposableHelper.DISPOSED);
       delegate.onComplete();
     }
   }
 
   @Override public void onError(Throwable e) {
     if (!isDisposed()) {
-      lazyDispose();
+      AutoDisposableHelper.dispose(lifecycleDisposable);
+      mainDisposable.lazySet(AutoDisposableHelper.DISPOSED);
```

Comment:
Swap these so the `mainDisposable` change gets flushed with the `dispose()` call.

Reply:
As in just swap the order they're executed?

```java
mainDisposable.lazySet(AutoDisposableHelper.DISPOSED);
AutoDisposableHelper.dispose(lifecycleDisposable);
```

Reply:
Assuming so after googling a bit - part of 9c3edf4